sparse/src/KokkosSparse_Utils.hpp

```diff
@@ -2330,6 +2330,163 @@ void validateCrsMatrix(int m, int n, const Rowmap &rowmapIn,
+  using size_type    = typename crsMat_t::non_const_size_type;
+  using offset_view1d_type =
+      Kokkos::View<size_type *, Kokkos::LayoutLeft, Kokkos::HostSpace>;
+
+  row_map_type A_row_map = A.graph.row_map;
+  entries_type A_entries = A.graph.entries;
+  values_type A_values   = A.values;
+
+  auto A_row_map_h =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), A_row_map);
+  auto A_entries_h =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), A_entries);
+  auto A_values_h =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace(), A_values);
+
+  ordinal_type A_nrows  = static_cast<ordinal_type>(A.numRows());
+  ordinal_type A_ncols  = static_cast<ordinal_type>(A.numCols());
+  ordinal_type n_blocks = static_cast<ordinal_type>(DiagBlk_v.size());
+
+  if (A_nrows != A_ncols) {
+    std::ostringstream os;
+    os << "The diagonal block extraction only works with square matrices -- "
+          "matrix A: "
+       << A_nrows << " x " << A_ncols;
+    throw std::runtime_error(os.str());
+  }
+
+  if (n_blocks == 1) {
+    // One block case: simply shallow copy A to DiagBlk_v[0]
+    DiagBlk_v[0] = crsMat_t(A);
+  } else {
+    // n_blocks > 1
+    if (A_nrows == 0) {
+      // Degenerate case: A is an empty matrix
+      for (ordinal_type i = 0; i < n_blocks; i++) {
+        DiagBlk_v[i] = crsMat_t();
+      }
+    } else {
+      // A_nrows >= 1
+      ordinal_type rows_per_block = ((A_nrows % n_blocks) == 0)
+                                        ? (A_nrows / n_blocks)
+                                        : (A_nrows / n_blocks + 1);
+
+      std::vector<out_row_map_type> row_map_v(n_blocks);
+      std::vector<out_entries_type> entries_v(n_blocks);
+      std::vector<out_values_type> values_v(n_blocks);
+      std::vector<out_row_map_hostmirror_type> row_map_h_v(n_blocks);
+      std::vector<out_entries_hostmirror_type> entries_h_v(n_blocks);
+      std::vector<out_values_hostmirror_type> values_h_v(n_blocks);
+
+      ordinal_type row_start = 0;  // first row index of i-th diagonal block
+      ordinal_type col_start = 0;  // first col index of i-th diagonal block
+      ordinal_type nrows, ncols;   // Nrows, Ncols of i-th diagonal block
+
+      for (ordinal_type i = 0; i < n_blocks; i++) {
+        nrows = rows_per_block;
+        if ((row_start + rows_per_block) > A_nrows) {
+          nrows = A_nrows - row_start;
+        }
+        col_start = row_start;
+        ncols     = nrows;
+
+        // Rowmap of i-th row-oriented sub-matrix
+        auto A_row_map_sub = Kokkos::subview(
+            A_row_map_h, Kokkos::make_pair(row_start, row_start + nrows + 1));
+
+        // First round: count i-th non-zeros or size of entries_v[i]
+        size_type n_entries = 0;
+        offset_view1d_type first("first", nrows);  // first position per row
+        offset_view1d_type last("last", nrows);    // last position per row
+
+        for (ordinal_type j = 0; j < nrows; j++) {  // loop through each row
```

Comment:
Consider breaking these sub-loops into some small functions, e.g. `find_start_col_idx`.

Reply:
@e10harvey Thanks. Although I think these sub-loops are only used in this utility function, I broke them into two small functions: ```kk_find_nnz_first_last_indices_subblock_crsmatrix_sequential``` and ```kk_extract_subblock_crsmatrix_sequential```